File: Ass_2.java
 import java.util.*;

public class Ass_2 {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        
        // Check if there is input to read
        if (sc.hasNextInt()) {
            int T = sc.nextInt(); // Number of test cases
            
            while (T-- > 0) {
                int N = sc.nextInt(); // Size of array
                int Q = sc.nextInt(); // Number of queries
                
                int[] products = new int[N];
                
                // Read the sorted product IDs
                for (int i = 0; i < N; i++) {
                    products[i] = sc.nextInt();
                }
                
                // Process each query
                while (Q-- > 0) {
                    int target = sc.nextInt();
                    
                    // Step 1: Find the first time the number appears
                    int firstIndex = findFirst(products, target);
                    
                    // If firstIndex is -1, the number is not in the list at all
                    if (firstIndex == -1) {
                        System.out.println(0);
                    } else {
                        // Step 2: Find the last time the number appears
                        int lastIndex = findLast(products, target);
                        
                        // Step 3: Calculate total count
                        int count = (lastIndex - firstIndex) + 1;
                        System.out.println(count);
                    }
                }
            }
        }
    }

    // Function to find the FIRST occurrence of target
    public static int findFirst(int[] arr, int target) {
        int start = 0;
        int end = arr.length - 1;
        int result = -1; // Store the potential answer here

        while (start <= end) {
            int mid = start + (end - start) / 2;

            if (arr[mid] == target) {
                result = mid;      // Found it! But is there one before it?
                end = mid - 1;     // Look to the LEFT
            } else if (target < arr[mid]) {
                end = mid - 1;
            } else {
                start = mid + 1;
            }
        }
        return result;
    }

    // Function to find the LAST occurrence of target
    public static int findLast(int[] arr, int target) {
        int start = 0;
        int end = arr.length - 1;
        int result = -1;

        while (start <= end) {
            int mid = start + (end - start) / 2;

            if (arr[mid] == target) {
                result = mid;      // Found it! But is there one after it?
                start = mid + 1;   // Look to the RIGHT
            } else if (target < arr[mid]) {
                end = mid - 1;
            } else {
                start = mid + 1;
            }
        }
        return result;
    }
}     

File: Ass2.py
import sys

def find_first_occurrence(arr, target):
    low = 0
    high = len(arr) - 1
    first_occurrence = -1
    while low <= high:
        mid = low + (high - low) // 2
        if arr[mid] == target:
            first_occurrence = mid
            high = mid - 1
        elif arr[mid] < target:
            low = mid + 1
        else:
            high = mid - 1
    return first_occurrence

def find_last_occurrence(arr, target):
    low = 0
    high = len(arr) - 1
    last_occurrence = -1
    while low <= high:
        mid = low + (high - low) // 2
        if arr[mid] == target:
            last_occurrence = mid
            low = mid + 1
        elif arr[mid] < target:
            low = mid + 1
        else:
            high = mid - 1
    return last_occurrence

def solve():
    # Prompt for N and Q
    line1 = input("\nEnter number of products (N) and number of queries (Q) separated by space: ").split()
    if not line1: return
    N, Q = map(int, line1)

    # Prompt for the list
    print(f"Enter {N} product IDs (sorted by size, e.g., 10 20 20 30):")
    product_ids = list(map(int, input().split()))

    for i in range(Q):
        query_id = int(input(f" Query {i+1}: Enter Product ID to search: "))
        first = find_first_occurrence(product_ids, query_id)
        
        if first == -1:
            print(f" Result: Product ID {query_id} found 0 times.")
        else:
            last = find_last_occurrence(product_ids, query_id)
            count = last - first + 1
            print(f" Result: Product ID {query_id} found {count} times.")

# Start of the program
try:
    T_str = input("Enter the number of test cases (T): ")
    T = int(T_str)
    for i in range(T):
        print(f"\n--- Test Case {i+1} ---")
        solve()
except ValueError:
    print("Please enter valid integers.")